applylib/applyset/applyset.go

```diff
@@ -153,11 +242,151 @@ func (a *ApplySet) ApplyOnce(ctx context.Context) (*ApplyResults, error) {
+		klog.V(4).Infof("Prune is enabled")
+		err := func() error {
+			pruneObjects, err := kapplyset.FindAllObjectsToPrune(ctx, a.client, visitedUids)
+			if err != nil {
+				return err
+			}
+			if err = a.deleteObjects(ctx, pruneObjects, results); err != nil {
+				return err
+			}
+			return nil
+		}()
+		if err != nil {
+			klog.Errorf("encounter error on pruning %v", err)
+			if e := a.updateParentLabelsAndAnnotations(ctx, kapplyset, "superset"); e != nil {
+				return results, e
+			}
+		}
+	}
+	if err := a.updateParentLabelsAndAnnotations(ctx, kapplyset, "latest"); err != nil {
+		klog.Errorf("update parent failed %v", err)
+	}
 	return results, nil
 }
+
+func mergeMap(from, to map[string]string) map[string]string {
+	if to == nil {
+		to = make(map[string]string)
+	}
+	for k, v := range from {
+		to[k] = v
+	}
+	return to
+}
+
+// updateLabel adds the "applyset.kubernetes.io/part-of: Parent-ID" label to the manifest.
+func (a *ApplySet) updateManifestLabel(obj ApplyableObject, applysetLabels map[string]string) error {
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("unable to convert `ApplyableObject` to `unstructured.Unstructured` %v/%v %v",
+			obj.GetName(), obj.GetNamespace(), obj.GroupVersionKind().String())
+	}
+	labels := u.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	newLabels := mergeMap(applysetLabels, labels)
+	u.SetLabels(newLabels)
+	return nil
+}
+
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+// This method leverages the kubectlapply to build the parent labels and annotations, but avoid using the
+// `resource.NewHelper` and cmdutil to patch the parent.
+func (a *ApplySet) updateParentLabelsAndAnnotations(ctx context.Context, kapplyset *kubectlapply.ApplySet, mode kubectlapply.ApplySetUpdateMode) error {
+	parent, err := meta.Accessor(a.parent.GetSubject())
+	if err != nil {
+		return err
+	}
+
+	original, err := meta.Accessor(a.parent.GetSubject().DeepCopyObject())
+	if err != nil {
+		return err
+	}
+	partialParent := kapplyset.BuildParentPatch(mode)
+	newAnnotations := mergeMap(partialParent.Annotations, parent.GetAnnotations())
+	parent.SetAnnotations(newAnnotations)
+	newLabels := mergeMap(partialParent.Labels, parent.GetLabels())
+	parent.SetLabels(newLabels)
+
+	return a.updateParent(ctx, original, parent)
+}
+
+func (a *ApplySet) deleteObjects(ctx context.Context, pruneObjects []kubectlapply.PruneObject, results *ApplyResults) error {
+	for i := range pruneObjects {
+		pruneObject := &pruneObjects[i]
+
+		name := pruneObject.Name
+		namespace := pruneObject.Namespace
+		mapping := pruneObject.Mapping
+		gvk := pruneObject.Object.GetObjectKind().GroupVersionKind()
+		nn := types.NamespacedName{Namespace: namespace, Name: name}
+
+		if err := a.client.Resource(mapping.Resource).Namespace(namespace).Delete(ctx, name, a.deleteOptions); err != nil {
+			klog.Error("unable to delete resource ")
```

Comment:
You might not need to log this at all, but if you do you should include `err`

Reply:
Sure, removed.